File: frontend/src/api/course.ts
import request from './index'
import type { AxiosPromise } from 'axios'

export interface Course {
  id: number
  name: string
  teacher: string
  semester: string
  description: string
  color: string
}

export interface CoursePayload {
  name: string
  teacher?: string
  semester?: string
  description?: string
  color?: string
}

export interface CourseListParams {
  page?: number
  page_size?: number
  keyword?: string
}

export interface CourseListResult {
  items: Course[]
  total: number
  page: number
  page_size: number
}

export function listCourses(params?: CourseListParams): AxiosPromise<CourseListResult> {
  return request.get('/courses', { params })
}

export function createCourse(payload: CoursePayload): AxiosPromise<Course> {
  return request.post('/courses', payload)
}

export function getCourse(id: number): AxiosPromise<Course> {
  return request.get(`/courses/${id}`)
}

export function updateCourse(id: number, payload: CoursePayload): AxiosPromise<Course> {
  return request.put(`/courses/${id}`, payload)
}

export function deleteCourse(id: number): AxiosPromise<void> {
  return request.delete(`/courses/${id}`)
}

File: backend/app/schemas/course.py
"""Pydantic schemas for the course endpoints."""
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class CourseBase(BaseModel):
    """Shared course fields."""

    name: str = Field(..., min_length=1, max_length=100)
    teacher: Optional[str] = Field(default=None, max_length=100)
    semester: Optional[str] = Field(default=None, max_length=50)
    description: Optional[str] = None
    color: Optional[str] = Field(default=None, max_length=20)


class CourseCreate(CourseBase):
    """Payload for POST /courses. ``user_id`` is bound server-side."""


class CourseUpdate(BaseModel):
    """Payload for PUT /courses/{id}. All fields optional."""

    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    teacher: Optional[str] = Field(default=None, max_length=100)
    semester: Optional[str] = Field(default=None, max_length=50)
    description: Optional[str] = None
    color: Optional[str] = Field(default=None, max_length=20)


class CourseResponse(CourseBase):
    """Course fields returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime


class CourseListResponse(BaseModel):
    """Paginated list of courses."""

    items: List[CourseResponse]
    total: int
    page: int
    page_size: int
